docs/apis/subsystems/roles.md

```diff
@@ -0,0 +1,612 @@
+
+1. `CAP_INHERIT`
+2. `CAP_ALLOW`
+3. `CAP_PREVENT`
+4. `CAP_PROHIBIT`
+
+If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.
+
+### Checking for capabilities
+
+You can use the [has_capability ()](https://docs.moodle.org/dev/Roles_and_modules#has_capability_.28.24capability.2C_.24contextid.2C_.24kill.29) function for checking for capabilities in your code.
+
+### Capability conflicts
+
+Since the capabilities in each role could be different, there could be conflict in capabilities. This is resolved using different methods in versions below Moodle 2.0.
+
+If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".
+
+### Capability-locality changes in v1.9
+
+When resolving conflicts between "allow" and "prevent" in v1.7 and v1.8 the locality of the capability is taken into account, although with less weight than the locality of the role assignment. In v1.9 the process has been reworked, we consider locality of the role assignment when dealing with allow/prevent conflicts, but ignore where the capability has been defined (as long as it applies to the context).
+
+Allow and prevent will cancel each other out if set for the same capability at the same context level. If this happens, we refer to the previous context level to determine the permission for the capability. This may sound more complex than it really is in practice.  The upshot is that the system can be flexible enough to allow pretty much any combination of permissions.
+
+For example - for the situation where there is
+
+- two roles assigned to one user in a context (eg student role and teacher role)
+- one override on one of those roles ( eg student role), at the same context
+
+then
+
+- **1.7/1.8** The override used to **always** win over the settings from the combined roles.
+
+- **1.9** The override is combined with the student role first. and then the roles are combined. This is more logical, as the override is acting like a modification to the student role.
+
+For more details, see [MDL-11218](http://tracker.moodle.org/browse/MDL-11218)
+
+### New permissions evaluation in 2.0
+
+The permission evaluation changes introduced in 1.9 were too complex for majority of uses, the permission evaluation was finally simplified in Moodle 2.0 - see [New permissions evaluation in 2.0](https://docs.moodle.org/dev/New_permissions_evaluation_in_2.0).
```

Comment:
Perhaps we should move this link to be our new ROles documentation since it's describing what we do with roles in 20+?

Reply:
Nothing on useful on New_permissions_evaluation_in_2.0, so just removing